contributors/design-proposals/controller-ref.md

```diff
@@ -1,100 +1,304 @@
-
-Secondary goal of this effort is to improve performance of various controllers and schedulers, by removing the need for expensive lookup for all matching "controllers".
-
-# Non goals
-
-Cascading deletion is not a goal of this effort. Cascading deletion will use `ownerReferences`, which is a [separate effort](garbage-collection.md).
-
-`ControllerRef` will extend `OwnerReference` and reuse machinery written for it (GarbageCollector, adoption/orphaning logic).
-
-# API and semantic changes
-
-There will be a new API field in the `OwnerReference` in which we will store an information if given owner is a managing controller:
-
-```
-OwnerReference {
-    …
-    Controller bool
+* [Goals](#goals)
+* [Non-goals](#non-goals)
+* [API](#api)
+* [Behavior](#behavior)
+* [Upgrading](#upgrading)
+* [Implementation](#implementation)
+* [Alternatives](#alternatives)
+
+# Goals
+
+* The main goal of ControllerRef (controller reference) is to solve the problem
+  of controllers that fight over controlled objects due to overlapping selectors
+  (e.g. a ReplicaSet fighting with a ReplicationController over Pods because
+  both controllers have label selectors that match those Pods).
+  Fighting controllers can [destabilize the apiserver](https://github.com/kubernetes/kubernetes/issues/24433),
+  [thrash objects back-and-forth](https://github.com/kubernetes/kubernetes/issues/24152),
+  or [cause controller operations to hang](https://github.com/kubernetes/kubernetes/issues/8598).
+
+  We don't want to have just an in-memory solution because we don't want a
+  Controller Manager crash to cause a massive reshuffling of controlled objects.
+  Therefore, we need to persist the mapping from each object to its controller.
+
+* A secondary goal of ControllerRef is to provide back-links from a given object
+  to the controller that manages it, which can be used for:
+  * Efficient object->controller lookup, without having to list all controllers.
+  * Generic object grouping (e.g. in a UI), without having to know about all
+    third-party controller types in advance.
+  * Replacing many if not all uses of the `kubernetes.io/created-by` annotation.
+
+# Non-goals
+
+* Overlapping selectors will continue to be considered user error.
+
+  ControllerRef will prevent this user error from destabilizing the cluster or
+  causing endless back-and-forth fighting between controllers, but it will not
+  make it completely safe to create controllers with overlapping selectors.
+
+  In particular, this proposal does not address cases such as Deployment or
+  StatefulSet, in which "families" of orphans may exist that ought to be adopted
+  as indivisible units.
+  Since multiple controllers may race to adopt orphans, the user must ensure
+  selectors do not overlap to avoid breaking up families.
+  Breaking up families of orphans could result in corruption or loss of
+  Deployment rollout state and history, and possibly also corruption or loss of
```

Comment:
In case of Deployments, what is the use case in which we want to leave the underlying ReplicaSet/pods around, and do a non-cascading deletion? How does that affect the rollout state/history?